core/state_processor.go

```diff
@@ -693,36 +693,102 @@ func (p *StateProcessor) Process(block *types.WorkObject, batch ethdb.Batch) (ty
+
+		targetBlock := p.hc.GetBlockByNumber(targetBlockNumber)
+
+		totalEntropy := big.NewInt(0)
+		target := new(big.Int).Div(common.Big2e256, targetBlock.Difficulty())
+		zoneThresholdEntropy := p.engine.IntrinsicLogEntropy(common.BytesToHash(target.Bytes()))
+		totalEntropy = new(big.Int).Add(totalEntropy, zoneThresholdEntropy)
+
+		// First step is to collect all the workshares and uncles at this targetBlockNumber depth, then
+		// compute the total entropy of the block, uncles and workshares at this level
+		// unclesAtTargetBlockDepth has all the uncles, workshares that is there at the block height
+		var sharesAtTargetBlockDepth []*types.WorkObjectHeader
+		var entropyOfSharesAtTargetBlockDepth []*big.Int
+		sharesAtTargetBlockDepth = append(sharesAtTargetBlockDepth, targetBlock.WorkObjectHeader())
+		entropyOfSharesAtTargetBlockDepth = append(entropyOfSharesAtTargetBlockDepth, zoneThresholdEntropy)
+
+		for i := 0; i <= params.WorkSharesInclusionDepth; i++ {
+			block := p.hc.GetBlockByNumber(targetBlockNumber + uint64(i))
+
+			var uncles []*types.WorkObjectHeader
+			if i == params.WorkSharesInclusionDepth {
+				uncles = targetBlock.Uncles()
+			} else {
+				uncles = block.Uncles()
+			}
+
+			for _, uncle := range uncles {
+				var uncleEntropy *big.Int
+				if uncle.NumberU64() == targetBlockNumber {
+					_, err := p.engine.VerifySeal(uncle)
+					if err != nil {
+						// uncle is a workshare
+						powHash, err := p.engine.ComputePowHash(uncle)
+						if err != nil {
+							return nil, nil, nil, nil, 0, 0, 0, nil, nil, err
+						}
+						uncleEntropy = new(big.Int).Set(p.engine.IntrinsicLogEntropy(powHash))
+						totalEntropy = new(big.Int).Add(totalEntropy, uncleEntropy)
+					} else {
+						// uncle is not a workshare
+						target := new(big.Int).Div(common.Big2e256, targetBlock.Difficulty())
+						uncleEntropy = new(big.Int).Set(p.engine.IntrinsicLogEntropy(common.BytesToHash(target.Bytes())))
+						totalEntropy = new(big.Int).Add(totalEntropy, zoneThresholdEntropy)
+					}
+					sharesAtTargetBlockDepth = append(sharesAtTargetBlockDepth, uncle)
+					entropyOfSharesAtTargetBlockDepth = append(entropyOfSharesAtTargetBlockDepth, uncleEntropy)
+				}
+			}
+		}
+
+		// Once the total entropy is calculated, the block reward is split
+		// between the blocks, uncles and workshares proportional to the block
+		// weight
+		parentOfTargetBlock := p.hc.GetBlockByHash(targetBlock.ParentHash(common.ZONE_CTX))
+
+		woHeader := &types.WorkObjectHeader{}
+		woHeader.SetPrimaryCoinbase(common.Bytes20ToAddress(common.ZeroInternal(common.Location{0, 0}), common.Location{0, 0}))
+		blockRewardAtTargetBlock := misc.CalculateReward(parentOfTargetBlock, woHeader)
+
+		// Add an etx for each workshare for it to be rewarded
+		for i, share := range sharesAtTargetBlockDepth {
+
+			shareReward := new(big.Int).Mul(blockRewardAtTargetBlock, entropyOfSharesAtTargetBlockDepth[i])
+			shareReward = new(big.Int).Div(shareReward, totalEntropy)
```

Comment:
We should add rationality checks on the divisor. Specifically totalEntropy > 0 at least. Additionally all shareReward < TotalEntropy

Reply:
done